collated/main/A0139498J.md

```diff
@@ -0,0 +1,1044 @@
+        assert targetIndexes != null;
+        this.targetIndexes = targetIndexes;
+    }
+
+    @Override
+    public CommandResult execute() {
+        
+        assert model != null;
+        // check with Edmund, can rename to isRedoAction()
+        if (!checkIfRedoAction()) {
+            setCurrentViewingList();
+        }
+        if (isViewingDoneList) {
+            logger.warning("Invalid command, cannot do done command on done list.");
+            indicateAttemptToExecuteIncorrectCommand();
+            return new CommandResult(String.format(Messages.MESSAGE_DONE_LIST_RESTRICTION));
+        }
+        
+        prepareToArchiveTasks(); 
+        archiveTasksFromGivenTargetIndexes();
+        updateHistory();
+        
+        if (doneTasks.isEmpty()) {
+            logger.warning("No tasks archived. Non of the given task indexes are valid.");
+            indicateAttemptToExecuteIncorrectCommand();
+            return new CommandResult(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+        }
+        String toDisplay = doneTasks.toString().replace("[", "").replace("]", "");
+        return (doneTasks.size() == 1)? 
+                new CommandResult(String.format(MESSAGE_DONE_ITEM_SUCCESS, toDisplay)):
+                new CommandResult(String.format(MESSAGE_DONE_ITEMS_SUCCESS, toDisplay));
+    }
+
+    /**
+     * Goes through the list of target indexes provided.
+     * Archives these tasks if target index is valid.
+     * Ignores any invalid target indexes provided.
+     */
+    private void archiveTasksFromGivenTargetIndexes() {
+        for (int targetIndex: targetIndexes) {
+            UnmodifiableObservableList<ReadOnlyTask> lastShownList = model.getFilteredUndoneTaskList();
+            boolean taskTargetIndexIsOutOfBounds = (lastShownList.size() < targetIndex - adjustmentForRemovedTask);
+            if (taskTargetIndexIsOutOfBounds) {
+                logger.warning("Task target index out of bounds detected. Skipping task with current target index.");
+                continue;
+            }
+            int adjustedTaskIndex = targetIndex - adjustmentForRemovedTask - 1;
+            Task taskToArchive = new Task(lastShownList.get(adjustedTaskIndex));
+            assert isViewingDoneList == false;
+            try {
+                model.deleteTask(taskToArchive);
+            } catch (TaskNotFoundException pnfe) {
+                assert false : "The target task cannot be missing";
+            }
+            doneTasks.add(taskToArchive);
+            if (taskToArchive.getRecurrenceRate().isPresent()) {
+                updateRecurrenceAndReAddTask(taskToArchive);
+            } else {
+                adjustmentForRemovedTask++;
+            }
+            model.addDoneTask(taskToArchive);
+        }
+    }
+
+    /**
+     * Adds the recurring task back into the undone task list
+     * with their start and end dates updated, if present.
+     * @param taskToArchive
```

Comment:
should highlight the significance of the parameter; else this line can be eliminated as it doesn't give any additional information